railties/lib/rails/engine.rb

```diff
@@ -396,6 +396,12 @@ def isolate_namespace(mod)
 
             unless mod.respond_to?(:table_name_prefix)
               define_method(:table_name_prefix) { "#{name}_" }
+
+              ActiveSupport.on_load(:active_record) do
+                mod.singleton_class.redefine_method(:table_name_prefix) do
+                  ActiveRecord::Base.table_name_prefix + "#{name}_"
+                end
+              end
             end
```

Comment:
> Doesn't this solution mean engines that do define their own table_name_prefix will still not obey config.active_record.table_name_prefix?

Initially I tried solving this in ActiveRecord, which worked but it broke some tests ( specifically these two https://github.com/rails/rails/blob/da871cc28a97af7bd4f410398d8d7fbed159b7c0/activerecord/test/cases/modules_test.rb#L109-L110)

And I didnt' want to introduce some potential breaking change in ActiveRecod for applications heavily using namespaces and relying on the current behavior.

==> That's why I opted to make this fix only for engines.

> Doesn't this solution mean engines that do define their own table_name_prefix will still not obey config.active_record.table_name_prefix?

Yes but I couldnt' figure out how to make it work both for engines that define their own. 
However, Maybe by preprending a module like so in an initializer?:
```ruby
initializer :active_record do
  ActiveSupport.on_load(:active_record) do
    module_eval do
      prepend(
        Module.new do
          def table_name_prefix
            ActiveRecord::Base.table_name_prefix + super
          end
       )
      end
    end 
  end
```

@jonathanhefner  I believe this solution will work for both `isolate_namespace` + those who have defined their own `table_name_prefix`.
What are your thoughts?

Reply:
Ok the initializer idea is a no go, I couldn't manage to make it work, because inside the engine I don't have access to the encapsulating module ( which isolate_namespace does thanks to the method argument)  :( 

Regarding `full_table_name_prefix`, the thing is I can't tell when iterating through `module_parents` which module encapsulates an engine, and which doesn't ( in order to not break existing behavior)

Or maybe even for regular namespaces ( module_parents) , `table_name_prefix` never really worked with `ActiveRecord#table_name_prefix`, I could need some help with a reproduction script

Reply:
Ok, I've been thinking this over, so here is a bit of a brain dump (some of it echoing what you've said)...

It didn't occur to me when I wrote https://github.com/rails/rails/pull/50167#issuecomment-1832448732, but this issue is not specific to `isolate_namespace`; it affects any namespace module that defines `table_name_prefix` _or_ `table_name_suffix`.  Here is another reproduction script:

<details>
  <summary>Reproduction script</summary>

  ```ruby
# frozen_string_literal: true

  require "bundler/inline"

  gemfile(true) do
    source "https://rubygems.org"

    git_source(:github) { |repo| "https://github.com/#{repo}.git" }

    gem "rails", github: "rails/rails", branch: "main"
    gem "sqlite3"
  end

  require "rails/all"

  module TestLib
    singleton_class.attr_accessor :table_name_prefix, :table_name_suffix
    self.table_name_prefix = "test_lib_"
    self.table_name_suffix = "_etc"

    class Record < ActiveRecord::Base
    end
  end

  class TestApp < Rails::Application
    config.load_defaults Rails::VERSION::STRING.to_f

    config.root = __dir__
    config.hosts << "example.org"
    config.eager_load = false
    config.session_store :cookie_store, key: "cookie_store_key"
    config.secret_key_base = "secret_key_base"

    config.logger = Logger.new($stdout)
    Rails.logger  = config.logger

    config.active_record.table_name_prefix = "abc_"
    config.active_record.table_name_suffix = "_xyz"
  end

  ENV["DATABASE_URL"] = "sqlite3::memory:"

  Rails.application.initialize!

  ActiveRecord::Schema.define do
    create_table :test_lib_records_etc do |t|
      t.string :value
    end
  end

  require "minitest/autorun"

  class LibTest < Minitest::Test
    def test_record
      TestLib::Record.create!(value: "foo")
      assert_equal "foo", TestLib::Record.first.value
    end
  end
  ```
</details>

Output:

```
(1.3ms)  CREATE TABLE "abc_test_lib_records_etc_xyz" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "value" varchar)

...

Error:
LibTest#test_record:
ActiveRecord::StatementInvalid: Could not find table 'test_lib_records_etc'
```

Basically, `table_name_{prefix,suffix}` module methods are incompatible with setting `config.active_record.table_name_{prefix,suffix}` because `create_table` will always apply `config.active_record.table_name_{prefix,suffix}` whereas `full_table_name_{prefix,suffix}` never will.

I'm not sure how to reconcile those two behaviors without breaking the tests you've linked.  So far, I've only come up with the following solutions:

1. Simply document that `table_name_{prefix,suffix}` module methods are incompatible with setting `config.active_record.table_name_{prefix,suffix}`.  However, this would mean users who set `config.active_record.table_name_{prefix,suffix}` would have to avoid libraries that define `table_name_{prefix,suffix}` module methods, including engines that use `isolated_namespace`.  And, as such, Rails would still need to address the problem in its own engines.

2. Change `create_table` to ignore `config.active_record.table_name_{prefix,suffix}` — i.e. the table name you specify is the table name you get.  This would be backward incompatible, but there is some leeway here thanks to migration versioning (e.g. using `ActiveRecord::Migration[7.2]`).  Though it would still be a significant and potentially upsetting change.

3. Add a `table_name_affixes: false` (or similar) option to `create_table` to disable applying `config.active_record.table_name_{prefix,suffix}`, plus document that it _must_ be specified when a `table_name_{prefix,suffix}` module method is defined concerning that table.  Libraries would potentially need to include the option in migrations they generate.

Unfortunately, all of those solutions involve some amount of inconvenience.

All that said, I believe the approach in this PR is compatible at least with solution <span>#</span>3.  Since the `table_name_prefix` method defined by `isolated_namespace` is a feature of Rails, I think it makes sense to integrate it with other Rails features, even if a "vanilla" `table_name_prefix` module method would not behave that way.  (And, to be clear, I don't think we need to implement any of the other solutions in this PR.)

Reply:
It would be really nice to implement this as an `attr_accessor` like:

```ruby
unless mod.respond_to?(:table_name_prefix)
  attr_accessor :table_name_prefix
  mod.table_name_prefix = "#{name}_"

  ActiveSupport.on_load(:active_record) do
    if mod.table_name_prefix == "#{name}_"
      mod.table_name_prefix = "#{table_name_prefix}#{name}_"
    end
  end
end
```

However, that doesn't work because the `on_load` block is called before `ActiveRecord::Base.table_name_prefix` is set.  We would need to set `mod.table_name_prefix` in some kind of callback that runs after the [`active_record.set_configs` initializer](https://github.com/rails/rails/blob/c4733f3bc4b04b7221af2bab35c3b18e566b251b/activerecord/lib/active_record/railtie.rb#L255-L301).

Still, a small tweak:

```suggestion
            unless mod.respond_to?(:table_name_prefix)
              define_method(:table_name_prefix) { "#{name}_" }

              ActiveSupport.on_load(:active_record) do
                mod.singleton_class.redefine_method(:table_name_prefix) do
                  "#{ActiveRecord::Base.table_name_prefix}#{name}_"
                end
              end
            end
```

Reply:
@jonathanhefner  Thank you for the reproduction script and thorough analysis, it is as I suspected, a bit complex.

If everything is ok with you, I think we can add, in another PR, some documentation about how engines that define their own `table_name_prefix` should make sure to add themselves the `ActiveRecord::Base.table_name_prefix`

(Maybe in the same PR that I intend to create for Action Text and Action Mailbox to remove all their hard-coded table names, or you can maybe do it in a subsequent PR / commit after merging this PR)

In any case, I feel we can merge both this PR, and this one https://github.com/rails/rails/pull/50167
and close both the related issues

Reply:
Sure, opening an issue would be good!  Hopefully, we can get a larger discussion going and consensus on the best approach. :smiley: